components/script/dom/console.rs

```diff
@@ -55,14 +64,138 @@ where
+                    let mut explicit_keys = object_class == ESClass::Object;
+                    let mut props = Vec::with_capacity(ids.len());
+                    for id in &*ids {
+                        rooted!(in(cx) let id = *id);
+                        rooted!(in(cx) let mut desc = PropertyDescriptor::default());
+
+                        let mut is_none = false;
+                        if !JS_GetOwnPropertyDescriptorById(
+                            cx,
+                            obj.handle().into(),
+                            id.handle().into(),
+                            desc.handle_mut().into(),
+                            &mut is_none,
+                        ) {
+                            return DOMString::from("/* invalid */");
+                        }
+
+                        rooted!(in(cx) let mut property = UndefinedValue());
+                        if !JS_GetPropertyById(cx, obj.handle(), id.handle(), property.handle_mut())
+                        {
+                            return DOMString::from("/* invalid */");
+                        }
+
+                        if !explicit_keys {
+                            if id.is_int() {
+                                if let Ok(id_int) = usize::try_from(id.to_int()) {
+                                    explicit_keys = props.len() != id_int;
+                                } else {
+                                    explicit_keys = false;
+                                }
+                            } else {
+                                explicit_keys = false;
+                            }
+                        }
+                        let value_string = stringify_inner(cx, property.handle(), parents.clone());
+                        if explicit_keys {
+                            let key = if id.is_string() || id.is_symbol() || id.is_int() {
+                                rooted!(in(cx) let mut key_value = UndefinedValue());
+                                let raw_id: jsapi::HandleId = id.handle().into();
+                                if !JS_IdToValue(cx, *raw_id.ptr, key_value.handle_mut()) {
+                                    return DOMString::from("/* invalid */");
+                                }
+                                value_to_source(cx, key_value.handle())
+                            } else {
+                                panic!("Key with impossible type")
+                            };
+                            props.push(format!("{}: {}", key, value_string,));
+                        } else {
+                            props.push(value_string.to_string());
+                        }
+                    }
+                    if object_class == ESClass::Array {
+                        DOMString::from(format!("[{}]", itertools::join(props, ", ")))
+                    } else {
+                        DOMString::from(format!("{{{}}}", itertools::join(props, ", ")))
+                    }
```

Comment:
This complex enough that perhaps there should be a `stringify_object_from_handle_value` function.